From c98bf1c700385d094f77f81b6fe285db6f9bca83 Mon Sep 17 00:00:00 2001
From: Jun Zhang <jun.zhang@arm.com>
Date: Fri, 28 Jul 2023 15:49:19 +0800
Subject: [PATCH] RSS: Enable Protected Storage service for kronos

To enable PS service, we need read/write the SI NVM
flash, but the old flash driver can not match the
requirement. Here replace the old flash driver to
CFI controller which is referenced from TC platform[0]
in Trusted-Firmware-M to access IntelStrataFlashJ3 flash.

The flash base for the PS: (SI_FLASH_BASE + 0x4000000)
The flash size for the PS: (0x10000)

0:
platform/ext/target/arm/rss/tc/host_drivers/cfi_drv.c
platform/ext/target/arm/rss/tc/host_drivers/cfi_drv.h
https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/

Upstream-Status: Pending
Signed-off-by: Jun Zhang <jun.zhang@arm.com>
---
 .../ext/target/arm/rss/common/CMakeLists.txt  |   5 +
 .../arm/rss/common/bl1/boot_hal_bl1_2.c       |   4 +-
 .../ext/target/arm/rss/common/config.cmake    |   3 +
 .../common/manifest/tfm_manifest_list.yaml    |  18 +-
 .../manifest/tfm_protected_storage.yaml       |  31 ++
 .../target/arm/rss/common/tfm_hal_platform.c  |  27 +-
 .../ext/target/arm/rss/kronos/CMakeLists.txt  |  25 ++
 .../target/arm/rss/kronos/bl2/boot_hal_bl2.c  |   8 +
 .../rss/kronos/cmsis_drivers/Driver_Flash.c   | 303 ++---------------
 .../cmsis_drivers/host_cmsis_driver_config.h  |   8 +
 .../ext/target/arm/rss/kronos/config.cmake    |   6 +-
 .../arm/rss/kronos/device/host_device_cfg.h   |   8 +-
 .../ext/target/arm/rss/kronos/flash_layout.h  |  40 ++-
 .../target/arm/rss/kronos/host_base_address.h |  31 +-
 .../arm/rss/kronos/host_drivers/cfi_drv.c     | 312 ++++++++++++++++++
 .../arm/rss/kronos/host_drivers/cfi_drv.h     |  85 +++++
 .../rss_comms/rss_comms_permissions_hal.c     |  18 +
 17 files changed, 648 insertions(+), 284 deletions(-)
 create mode 100644 platform/ext/target/arm/rss/common/manifest/tfm_protected_storage.yaml
 create mode 100644 platform/ext/target/arm/rss/kronos/host_drivers/cfi_drv.c
 create mode 100644 platform/ext/target/arm/rss/kronos/host_drivers/cfi_drv.h

diff --git a/platform/ext/target/arm/rss/common/CMakeLists.txt b/platform/ext/target/arm/rss/common/CMakeLists.txt
index d905ad068..cb859f554 100644
--- a/platform/ext/target/arm/rss/common/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/common/CMakeLists.txt
@@ -373,6 +373,11 @@ target_sources(platform_bl1_2
 )
 
 target_include_directories(platform_bl1_2
+    PUBLIC
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/cfi
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
+        $<$<NOT:$<BOOL:${RSS_DEBUG_UART}>>:${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011>
     PRIVATE
         .
 )
diff --git a/platform/ext/target/arm/rss/common/bl1/boot_hal_bl1_2.c b/platform/ext/target/arm/rss/common/bl1/boot_hal_bl1_2.c
index d2474a5cb..5b5d4d54a 100644
--- a/platform/ext/target/arm/rss/common/bl1/boot_hal_bl1_2.c
+++ b/platform/ext/target/arm/rss/common/bl1/boot_hal_bl1_2.c
@@ -129,9 +129,9 @@ int32_t boot_platform_post_init(void)
     enum atu_error_t err;
 
     /* Initialize SI NVM flash ATU region */
-    /* Allocate 128MiB for SI NVM flash as the limit of per ATU region size */
+    /* Allocate 64MiB for SI NVM flash as the limit of per ATU region size */
     err = atu_initialize_region(&ATU_DEV_S, RSS_ATU_SI_FLASH_ID, SI_FLASH_BASE_NS_LOG,
-                               SI_FLASH_BASE_NS_PHY, SI_FLASH_SIZE);
+                               SI_FLASH_BASE_NS_PHY, SI_FLASH_IMG_SIZE);
 
     if (err != ATU_ERR_NONE) {
         return 1;
diff --git a/platform/ext/target/arm/rss/common/config.cmake b/platform/ext/target/arm/rss/common/config.cmake
index e222d6b5f..d4dcedcf9 100644
--- a/platform/ext/target/arm/rss/common/config.cmake
+++ b/platform/ext/target/arm/rss/common/config.cmake
@@ -74,6 +74,9 @@ set(TEST_NS_MULTI_CORE                  OFF        CACHE BOOL     "Whether to bu
 configure_file(${CMAKE_CURRENT_LIST_DIR}/manifest/tfm_manifest_list.yaml ${CMAKE_BINARY_DIR}/tools/tfm_manifest_list.yaml)
 set(TFM_MANIFEST_LIST                   ${CMAKE_BINARY_DIR}/tools/tfm_manifest_list.yaml CACHE FILEPATH "TF-M native Secure Partition manifests list file")
 
+set(TFM_PLAT_SPECIFIC_MANIFEST_FILE "${CMAKE_CURRENT_LIST_DIR}/manifest/tfm_manifest_list.yaml" CACHE PATH "Platform specific Secure Partition manifests file")
+configure_file(${TFM_PLAT_SPECIFIC_MANIFEST_FILE} ${CMAKE_BINARY_DIR}/tools/tfm_manifest_list.yaml)
+
 # Platform-specific configurations
 set(CONFIG_TFM_USE_TRUSTZONE            OFF)
 set(TFM_MULTI_CORE_TOPOLOGY             ON)
diff --git a/platform/ext/target/arm/rss/common/manifest/tfm_manifest_list.yaml b/platform/ext/target/arm/rss/common/manifest/tfm_manifest_list.yaml
index 14a6ab1a3..d02870b21 100644
--- a/platform/ext/target/arm/rss/common/manifest/tfm_manifest_list.yaml
+++ b/platform/ext/target/arm/rss/common/manifest/tfm_manifest_list.yaml
@@ -1,6 +1,6 @@
 #-------------------------------------------------------------------------------
-# Copyright (c) 2018-2022, Arm Limited. All rights reserved.
-# Copyright (c) 2021-2022 Cypress Semiconductor Corporation (an Infineon company)
+# Copyright (c) 2018-2023, Arm Limited. All rights reserved.
+# Copyright (c) 2021-2023 Cypress Semiconductor Corporation (an Infineon company)
 # or an affiliate of Cypress Semiconductor Corporation. All rights reserved.
 #
 # SPDX-License-Identifier: BSD-3-Clause
@@ -37,6 +37,20 @@
       },
       "non_ffm_attributes": ['ns_agent']
     },
+    {
+      "description": "Protected Storage Partition",
+      "manifest": "${CMAKE_SOURCE_DIR}/platform/ext/target/arm/rss/common/manifest/tfm_protected_storage.yaml",
+      "output_path": "secure_fw/partitions/protected_storage",
+      "conditional": "TFM_PARTITION_PROTECTED_STORAGE",
+      "version_major": 0,
+      "version_minor": 1,
+      "pid": 256,
+      "linker_pattern": {
+        "library_list": [
+           "*tfm_*partition_ps.*"
+         ],
+      }
+    },
     {
       "description": "TFM Crypto Service",
       "manifest": "${CMAKE_SOURCE_DIR}/platform/ext/target/arm/rss/common/manifest/tfm_crypto.yaml",
diff --git a/platform/ext/target/arm/rss/common/manifest/tfm_protected_storage.yaml b/platform/ext/target/arm/rss/common/manifest/tfm_protected_storage.yaml
new file mode 100644
index 000000000..859e77439
--- /dev/null
+++ b/platform/ext/target/arm/rss/common/manifest/tfm_protected_storage.yaml
@@ -0,0 +1,31 @@
+#-------------------------------------------------------------------------------
+# Copyright (c) 2018-2023, Arm Limited. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+#-------------------------------------------------------------------------------
+
+{
+  "psa_framework_version": 1.1,
+  "name": "TFM_SP_PS",
+  "type": "APPLICATION-ROT",
+  "priority": "NORMAL",
+  "model": "SFN",
+  "entry_init": "tfm_ps_entry",
+  "stack_size": "PS_STACK_SIZE",
+  "services" : [
+    {
+      "name": "TFM_PROTECTED_STORAGE_SERVICE",
+      "sid": "0x00000060",
+      "non_secure_clients": true,
+      "connection_based": false,
+      "stateless_handle": 2,
+      "version": 1,
+      "version_policy": "STRICT"
+    }
+  ],
+  "dependencies": [
+    "TFM_CRYPTO",
+    "TFM_PLATFORM_SERVICE"
+  ]
+}
diff --git a/platform/ext/target/arm/rss/common/tfm_hal_platform.c b/platform/ext/target/arm/rss/common/tfm_hal_platform.c
index b81d98a77..5b975597a 100644
--- a/platform/ext/target/arm/rss/common/tfm_hal_platform.c
+++ b/platform/ext/target/arm/rss/common/tfm_hal_platform.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021-2022, Arm Limited. All rights reserved.
+ * Copyright (c) 2021-2023, Arm Limited. All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  *
@@ -12,11 +12,23 @@
 #include "uart_stdout.h"
 #include "device_definition.h"
 
+#ifdef TFM_PARTITION_PROTECTED_STORAGE
+#include "host_base_address.h"
+#include "size_defs.h"
+
+#define  RSS_ATU_REGION_PS_SLOT  16
+#endif /* TFM_PARTITION_PROTECTED_STORAGE */
+
+
 extern const struct memory_region_limits memory_regions;
 
 enum tfm_hal_status_t tfm_hal_platform_init(void)
 {
     enum tfm_plat_err_t plat_err = TFM_PLAT_ERR_SYSTEM_ERR;
+#ifdef TFM_PARTITION_PROTECTED_STORAGE
+    enum atu_error_t err;
+#endif /* TFM_PARTITION_PROTECTED_STORAGE */
+
 
     plat_err = enable_fault_handlers();
     if (plat_err != TFM_PLAT_ERR_SUCCESS) {
@@ -51,6 +63,19 @@ enum tfm_hal_status_t tfm_hal_platform_init(void)
         return TFM_HAL_ERROR_GENERIC;
     }
 
+#ifdef TFM_PARTITION_PROTECTED_STORAGE
+    /* Initialize PS region */
+    err = atu_initialize_region(&ATU_DEV_S,
+                                RSS_ATU_REGION_PS_SLOT,
+                                HOST_ACCESS_PS_BASE_S,
+                                HOST_FLASH0_PS_BASE,
+                                HOST_FLASH0_PS_SIZE);
+    if (err != ATU_ERR_NONE) {
+        return TFM_HAL_ERROR_GENERIC;
+    }
+
+#endif /* TFM_PARTITION_PROTECTED_STORAGE */
+
     return TFM_HAL_SUCCESS;
 }
 
diff --git a/platform/ext/target/arm/rss/kronos/CMakeLists.txt b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
index bb6fb5abd..9c768a739 100644
--- a/platform/ext/target/arm/rss/kronos/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
@@ -35,12 +35,16 @@ target_include_directories(tfm_spm
 target_include_directories(platform_s
     PUBLIC
         cmsis_drivers
+        host_drivers
         device
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
 )
 
 target_sources(platform_s
     PRIVATE
+        host_drivers/cfi_drv.c
         cmsis_drivers/Driver_Flash.c
         cmsis_drivers/Driver_USART.c
         device/host_device_definition.c
@@ -50,6 +54,7 @@ target_sources(platform_s
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011/uart_pl011_drv.c
         rss_expansion_peripherals_def.c
         rss_comms/rss_comms_permissions_hal.c
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata/spi_strataflashj3_flash_lib.c
 )
 
 target_sources(tfm_s
@@ -57,11 +62,14 @@ target_sources(tfm_s
         device/rss_expansion_device_definition.c
         ${CMAKE_CURRENT_LIST_DIR}/../common/mscp_atu_request_handler.c
         ${CMAKE_CURRENT_LIST_DIR}/../common/mscp_mailbox.c
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata/spi_strataflashj3_flash_lib.c
 )
 
 target_include_directories(tfm_s
     PUBLIC
         ./device
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
 )
 
 #========================= Platform Non-Secure ================================#
@@ -70,16 +78,21 @@ target_include_directories(platform_ns
     PUBLIC
         cmsis_drivers
         device
+        host_drivers
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
 )
 
 target_sources(platform_ns
     PRIVATE
+        host_drivers/cfi_drv.c
         cmsis_drivers/Driver_Flash.c
         cmsis_drivers/Driver_USART.c
         device/host_device_definition.c
         device/rss_expansion_device_definition.c
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011/uart_pl011_drv.c
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata/spi_strataflashj3_flash_lib.c
 )
 
 #========================= Platform BL2 =======================================#
@@ -91,10 +104,12 @@ target_sources(platform_bl2
         cmsis_drivers/Driver_USART.c
         device/host_device_definition.c
         device/rss_expansion_device_definition.c
+        host_drivers/cfi_drv.c
         ${CMAKE_CURRENT_LIST_DIR}/../common/mscp_atu_request_handler.c
         ${CMAKE_CURRENT_LIST_DIR}/../common/mscp_mailbox.c
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011/uart_pl011_drv.c
         tfm_interrupts.c
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata/spi_strataflashj3_flash_lib.c
 )
 
 target_compile_definitions(platform_bl2
@@ -107,13 +122,18 @@ target_sources(bl2
     PRIVATE
         ${CMAKE_CURRENT_LIST_DIR}/../common/mscp_atu_request_handler.c
         ${CMAKE_CURRENT_LIST_DIR}/../common/mscp_mailbox.c
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
 )
 
 target_include_directories(platform_bl2
     PUBLIC
         cmsis_drivers
         device
+        host_drivers
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
 )
 
 target_compile_definitions(platform_bl2
@@ -125,18 +145,23 @@ target_compile_definitions(platform_bl2
 
 target_sources(platform_bl1_1
     PRIVATE
+        host_drivers/cfi_drv.c
         cmsis_drivers/Driver_Flash.c
         cmsis_drivers/Driver_USART.c
         device/host_device_definition.c
         device/rss_expansion_device_definition.c
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011/uart_pl011_drv.c
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata/spi_strataflashj3_flash_lib.c
 )
 
 target_include_directories(platform_bl1_1_interface
     INTERFACE
         cmsis_drivers
         device
+        host_drivers
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
+        ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
 )
 
 #=========================== MHUv3 Dependent Code ===============================#
diff --git a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
index 1b73e21fd..37d402d6c 100644
--- a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
+++ b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
@@ -33,6 +33,7 @@
 #endif /* CRYPTO_HW_ACCELERATOR */
 
 extern volatile bool scp_doorbell;
+extern ARM_DRIVER_FLASH FLASH_DEV_NAME;
 
 static uint8_t *lcp_measurement;
 static struct boot_measurement_metadata *lcp_measurement_metadata;
@@ -200,8 +201,15 @@ int boot_store_measurement(uint8_t index,
 
 int32_t boot_platform_init(void)
 {
+    int32_t result;
+
     read_chip_id();
 
+    result = FLASH_DEV_NAME.Initialize(NULL);
+    if (result != ARM_DRIVER_OK) {
+        return 1;
+    }
+
     return 0;
 }
 
diff --git a/platform/ext/target/arm/rss/kronos/cmsis_drivers/Driver_Flash.c b/platform/ext/target/arm/rss/kronos/cmsis_drivers/Driver_Flash.c
index 8bbf82778..aa06b2801 100644
--- a/platform/ext/target/arm/rss/kronos/cmsis_drivers/Driver_Flash.c
+++ b/platform/ext/target/arm/rss/kronos/cmsis_drivers/Driver_Flash.c
@@ -18,10 +18,11 @@
 
 #include <string.h>
 #include <stdint.h>
-#include "Driver_Flash.h"
-#include "RTE_Device.h"
 #include "platform_base_address.h"
 #include "host_base_address.h"
+#include "Driver_Flash_Strata.h"
+#include "cmsis_driver_config.h"
+#include "RTE_Device.h"
 
 #ifndef ARG_UNUSED
 #define ARG_UNUSED(arg)  ((void)arg)
@@ -33,283 +34,43 @@
 #define FLASH0_PAGE_SIZE      0x00001000 /* 4 kB */
 #define FLASH0_PROGRAM_UNIT   0x1        /* Minimum write size */
 
-/* Driver version */
-#define ARM_FLASH_DRV_VERSION      ARM_DRIVER_VERSION_MAJOR_MINOR(1, 1)
-#define ARM_FLASH_DRV_ERASE_VALUE  0xFF
-
-/**
- * Data width values for ARM_FLASH_CAPABILITIES::data_width
- * \ref ARM_FLASH_CAPABILITIES
- */
- enum {
-    DATA_WIDTH_8BIT   = 0u,
-    DATA_WIDTH_16BIT,
-    DATA_WIDTH_32BIT,
-    DATA_WIDTH_ENUM_SIZE
+#if (RTE_FLASH0)
+static ARM_FLASH_INFO ARM_FLASH0_DEV_DATA = {
+    .sector_info    = NULL,     /* Uniform sector layout */
+    .sector_count   = HOST_ACCESS_SIZE / 0x1000,
+    .sector_size    = 0x1000,
+    .page_size      = 256U,
+    .program_unit   = 1U,
+    .erased_value   = ARM_FLASH_DRV_ERASE_VALUE
 };
 
-static const uint32_t data_width_byte[DATA_WIDTH_ENUM_SIZE] = {
-    sizeof(uint8_t),
-    sizeof(uint16_t),
-    sizeof(uint32_t),
+static struct arm_strata_flash_dev_t ARM_FLASH0_DEV = {
+    .dev    = &FLASH0_DEV,
+    .data   = &ARM_FLASH0_DEV_DATA
 };
 
-/*
- * ARM FLASH device structure
- *
- * This driver just emulates a flash interface and behaviour on top of the SRAM
- * memory.
- */
-struct arm_flash_dev_t {
-    const uint32_t memory_base;   /*!< FLASH memory base address */
-    ARM_FLASH_INFO *data;         /*!< FLASH data */
-};
+ARM_FLASH_STRATA(ARM_FLASH0_DEV, Driver_FLASH0);
 
-/* Flash Status */
-static ARM_FLASH_STATUS FlashStatus = {0, 0, 0};
-
-/* Driver Version */
-static const ARM_DRIVER_VERSION DriverVersion = {
-    ARM_FLASH_API_VERSION,
-    ARM_FLASH_DRV_VERSION
+#if (defined (SPI_STRATAFLASHJ3_S) && defined (CFI_S))
+static const struct cfi_dev_cfg_t CFI_DEV_CFG_S = {
+    /* Define the flash base/size to be the same as the host access area, as the
+     * flash may not be mapped contiguously or predictably within that area.
+     */
+    .base = SI_FLASH_BASE_NS_LOG,
 };
 
-/* Driver Capabilities */
-static const ARM_FLASH_CAPABILITIES DriverCapabilities = {
-    0, /* event_ready */
-    0, /* data_width = 0:8-bit, 1:16-bit, 2:32-bit */
-    1  /* erase_chip */
+struct cfi_dev_t CFI_DEV_S = {
+    .cfg = &CFI_DEV_CFG_S,
 };
 
-static int32_t is_range_valid(struct arm_flash_dev_t *flash_dev,
-                              uint32_t offset)
-{
-    uint32_t flash_limit = 0;
-    int32_t rc = 0;
-
-    flash_limit = (flash_dev->data->sector_count * flash_dev->data->sector_size)
-                   - 1;
-
-    if (offset > flash_limit) {
-        rc = -1;
-    }
-    return rc;
-}
-
-static int32_t is_write_aligned(struct arm_flash_dev_t *flash_dev,
-                                uint32_t param)
-{
-    int32_t rc = 0;
-
-    if ((param % flash_dev->data->program_unit) != 0) {
-        rc = -1;
-    }
-    return rc;
-}
-
-static int32_t is_sector_aligned(struct arm_flash_dev_t *flash_dev,
-                                 uint32_t offset)
-{
-    int32_t rc = 0;
-
-    if ((offset % flash_dev->data->sector_size) != 0) {
-        rc = -1;
-    }
-    return rc;
-}
-
-static int32_t is_flash_ready_to_write(const uint8_t *start_addr, uint32_t cnt)
-{
-    int32_t rc = 0;
-    uint32_t i;
-
-    for (i = 0; i < cnt; i++) {
-        if(start_addr[i] != ARM_FLASH_DRV_ERASE_VALUE) {
-            rc = -1;
-            break;
-        }
-    }
-
-    return rc;
-}
-
-#if (RTE_FLASH0)
-static ARM_FLASH_INFO ARM_FLASH0_DEV_DATA = {
-    .sector_info  = NULL,                  /* Uniform sector layout */
-    .sector_count = FLASH0_SIZE / FLASH0_SECTOR_SIZE,
-    .sector_size  = FLASH0_SECTOR_SIZE,
-    .page_size    = FLASH0_PAGE_SIZE,
-    .program_unit = FLASH0_PROGRAM_UNIT,
-    .erased_value = ARM_FLASH_DRV_ERASE_VALUE};
-
-static struct arm_flash_dev_t ARM_FLASH0_DEV = {
-    .memory_base = SI_FLASH_BASE_NS_LOG,
-    .data        = &(ARM_FLASH0_DEV_DATA)};
-
-struct arm_flash_dev_t *FLASH0_DEV = &ARM_FLASH0_DEV;
-
-/*
- * Functions
- */
-
-static ARM_DRIVER_VERSION ARM_Flash_GetVersion(void)
-{
-    return DriverVersion;
-}
-
-static ARM_FLASH_CAPABILITIES ARM_Flash_GetCapabilities(void)
-{
-    return DriverCapabilities;
-}
-
-static int32_t ARM_Flash_Initialize(ARM_Flash_SignalEvent_t cb_event)
-{
-    ARG_UNUSED(cb_event);
-
-    if (DriverCapabilities.data_width >= DATA_WIDTH_ENUM_SIZE) {
-        return ARM_DRIVER_ERROR;
-    }
-
-    /* Nothing to be done */
-    return ARM_DRIVER_OK;
-}
-
-static int32_t ARM_Flash_Uninitialize(void)
-{
-    /* Nothing to be done */
-    return ARM_DRIVER_OK;
-}
-
-static int32_t ARM_Flash_PowerControl(ARM_POWER_STATE state)
-{
-    switch (state) {
-    case ARM_POWER_FULL:
-        /* Nothing to be done */
-        return ARM_DRIVER_OK;
-        break;
-
-    case ARM_POWER_OFF:
-    case ARM_POWER_LOW:
-    default:
-        return ARM_DRIVER_ERROR_UNSUPPORTED;
-    }
-}
-
-static int32_t ARM_Flash_ReadData(uint32_t addr, void *data, uint32_t cnt)
-{
-    uint32_t start_addr = FLASH0_DEV->memory_base + addr;
-    int32_t rc = 0;
-
-    /* Conversion between data items and bytes */
-    cnt *= data_width_byte[DriverCapabilities.data_width];
-
-    /* Check flash memory boundaries */
-    rc = is_range_valid(FLASH0_DEV, addr + cnt);
-    if (rc != 0) {
-        return ARM_DRIVER_ERROR_PARAMETER;
-    }
-
-    /* Flash interface just emulated over SRAM, use memcpy */
-    memcpy(data, (void *)start_addr, cnt);
-
-    /* Conversion between bytes and data items */
-    cnt /= data_width_byte[DriverCapabilities.data_width];
-
-    return cnt;
-}
-
-static int32_t ARM_Flash_ProgramData(uint32_t addr, const void *data,
-                                     uint32_t cnt)
-{
-    uint32_t start_addr = FLASH0_DEV->memory_base + addr;
-    int32_t rc = 0;
-
-    /* Conversion between data items and bytes */
-    cnt *= data_width_byte[DriverCapabilities.data_width];
-
-    /* Check flash memory boundaries and alignment with minimal write size */
-    rc  = is_range_valid(FLASH0_DEV, addr + cnt);
-    rc |= is_write_aligned(FLASH0_DEV, addr);
-    rc |= is_write_aligned(FLASH0_DEV, cnt);
-    if (rc != 0) {
-        return ARM_DRIVER_ERROR_PARAMETER;
-    }
-
-    /* Check if the flash area to write the data was erased previously */
-    rc = is_flash_ready_to_write((const uint8_t*)start_addr, cnt);
-    if (rc != 0) {
-        return ARM_DRIVER_ERROR;
-    }
-
-    /* Flash interface just emulated over SRAM, use memcpy */
-    memcpy((void *)start_addr, data, cnt);
-
-    /* Conversion between bytes and data items */
-    cnt /= data_width_byte[DriverCapabilities.data_width];
-
-    return cnt;
-}
-
-static int32_t ARM_Flash_EraseSector(uint32_t addr)
-{
-    uint32_t start_addr = FLASH0_DEV->memory_base + addr;
-    uint32_t rc = 0;
-
-    rc  = is_range_valid(FLASH0_DEV, addr);
-    rc |= is_sector_aligned(FLASH0_DEV, addr);
-    if (rc != 0) {
-        return ARM_DRIVER_ERROR_PARAMETER;
-    }
-
-    /* Flash interface just emulated over SRAM, use memset */
-    memset((void *)start_addr,
-           FLASH0_DEV->data->erased_value,
-           FLASH0_DEV->data->sector_size);
-    return ARM_DRIVER_OK;
-}
-
-static int32_t ARM_Flash_EraseChip(void)
-{
-    uint32_t i;
-    uint32_t addr = FLASH0_DEV->memory_base;
-    int32_t rc = ARM_DRIVER_ERROR_UNSUPPORTED;
-
-    /* Check driver capability erase_chip bit */
-    if (DriverCapabilities.erase_chip == 1) {
-        for (i = 0; i < FLASH0_DEV->data->sector_count; i++) {
-            /* Flash interface just emulated over SRAM, use memset */
-            memset((void *)addr,
-                   FLASH0_DEV->data->erased_value,
-                   FLASH0_DEV->data->sector_size);
-
-            addr += FLASH0_DEV->data->sector_size;
-            rc = ARM_DRIVER_OK;
-        }
-    }
-    return rc;
-}
-
-static ARM_FLASH_STATUS ARM_Flash_GetStatus(void)
-{
-    return FlashStatus;
-}
-
-static ARM_FLASH_INFO * ARM_Flash_GetInfo(void)
-{
-    return FLASH0_DEV->data;
-}
-
-ARM_DRIVER_FLASH Driver_FLASH0 = {
-    ARM_Flash_GetVersion,
-    ARM_Flash_GetCapabilities,
-    ARM_Flash_Initialize,
-    ARM_Flash_Uninitialize,
-    ARM_Flash_PowerControl,
-    ARM_Flash_ReadData,
-    ARM_Flash_ProgramData,
-    ARM_Flash_EraseSector,
-    ARM_Flash_EraseChip,
-    ARM_Flash_GetStatus,
-    ARM_Flash_GetInfo
+struct cfi_strataflashj3_dev_t SPI_STRATAFLASHJ3_DEV = {
+    .controller = &CFI_DEV_S,
+    .total_sector_cnt = 0,
+    .page_size = 0,
+    .sector_size = 0,
+    .program_unit = 1U,
+    .is_initialized = false,
 };
+#endif /* (SPI_STRATAFLASHJ3_S) && (CFI_S) */
+
 #endif /* RTE_FLASH0 */
diff --git a/platform/ext/target/arm/rss/kronos/cmsis_drivers/host_cmsis_driver_config.h b/platform/ext/target/arm/rss/kronos/cmsis_drivers/host_cmsis_driver_config.h
index d26b9d842..2e4d0136c 100644
--- a/platform/ext/target/arm/rss/kronos/cmsis_drivers/host_cmsis_driver_config.h
+++ b/platform/ext/target/arm/rss/kronos/cmsis_drivers/host_cmsis_driver_config.h
@@ -19,4 +19,12 @@
 
 /* Currently no host cmsis drivers */
 
+#include "cmsis.h"
+#include "device_definition.h"
+#include "RTE_Device.h"
+
+extern struct cfi_strataflashj3_dev_t SPI_STRATAFLASHJ3_DEV;
+
+#define FLASH0_DEV                   SPI_STRATAFLASHJ3_DEV
+
 #endif  /* __HOST_CMSIS_DRIVER_CONFIG_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/config.cmake b/platform/ext/target/arm/rss/kronos/config.cmake
index 5da95c7c5..054116161 100644
--- a/platform/ext/target/arm/rss/kronos/config.cmake
+++ b/platform/ext/target/arm/rss/kronos/config.cmake
@@ -27,7 +27,11 @@ if(MCUBOOT_IMAGE_NUMBER GREATER 9)
     message(FATAL_ERROR "INVALID CONFIG: MCUBOOT_IMAGE_NUMBER value not supported!")
 endif()
 
-set(PLATFORM_HAS_PS_NV_OTP_COUNTERS            ON  CACHE BOOL   "Platform supports nvm counters for PS in OTP")
+set(PLATFORM_HAS_PS_NV_OTP_COUNTERS           ON   CACHE BOOL   "Platform supports nvm counters for PS in OTP")
+set(TFM_PARTITION_MEASURED_BOOT               OFF  CACHE BOOL   "Disable Measured boot partition")
+set(TFM_PARTITION_DELEGATED_ATTESTATION       OFF  CACHE BOOL   "Disable Delegated Attestation partition")
+set(TFM_PARTITION_PROTECTED_STORAGE           ON   CACHE BOOL   "Enable Protected Storage partition")
+set(TFM_PARTITION_INTERNAL_TRUSTED_STORAGE    OFF  CACHE BOOL   "Disable Internal Trusted Storage partition")
 
 # Once all options are set, set common options as fallback
 include(${CMAKE_CURRENT_LIST_DIR}/../common/config.cmake)
diff --git a/platform/ext/target/arm/rss/kronos/device/host_device_cfg.h b/platform/ext/target/arm/rss/kronos/device/host_device_cfg.h
index ecf66a238..c6142a131 100644
--- a/platform/ext/target/arm/rss/kronos/device/host_device_cfg.h
+++ b/platform/ext/target/arm/rss/kronos/device/host_device_cfg.h
@@ -27,4 +27,10 @@
 #define MHU_V3_RSS_TO_SCP
 #define MHU_V3_SCP_TO_RSS
 
-#endif /* __HOST_DEVICE_CFG_H__ */
\ No newline at end of file
+/*External Intel Strata Flash Device*/
+#define SPI_STRATAFLASHJ3_S
+
+/* CFI Controller */
+#define CFI_S
+
+#endif /* __HOST_DEVICE_CFG_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/flash_layout.h b/platform/ext/target/arm/rss/kronos/flash_layout.h
index 078448c20..2885e97a9 100644
--- a/platform/ext/target/arm/rss/kronos/flash_layout.h
+++ b/platform/ext/target/arm/rss/kronos/flash_layout.h
@@ -64,7 +64,7 @@
 /* Sector size of the flash hardware; same as FLASH0_SECTOR_SIZE */
 #define FLASH_AREA_IMAGE_SECTOR_SIZE    (0x1000)    /* 4 KB */
 /* Same as FLASH0_SIZE */
-#define FLASH_TOTAL_SIZE                (SI_FLASH_SIZE)  /* 128 MB */
+#define FLASH_TOTAL_SIZE                (SI_FLASH_IMG_SIZE)  /* 64 MB */
 
 /* Flash layout info for BL2 bootloader */
 /* Same as SI_FLASH_BASE_NS_LOG */
@@ -207,4 +207,42 @@
 /* Smallest flash programmable unit in bytes */
 #define TFM_HAL_FLASH_PROGRAM_UNIT      (0x1)
 
+#ifdef TFM_PARTITION_PROTECTED_STORAGE
+/*******************************/
+/* ITS */
+/*******************************/
+#define FLASH_DEV_NAME_SE_SECURE_FLASH  Driver_FLASH0
+
+/* Internal Trusted Storage (ITS) Service definitions
+ * Note: Further documentation of these definitions can be found in the
+ * TF-M ITS Integration Guide.
+ */
+#define TFM_HAL_ITS_FLASH_DRIVER FLASH_DEV_NAME_SE_SECURE_FLASH
+
+/* Smallest flash programmable unit in bytes */
+#define TFM_HAL_ITS_PROGRAM_UNIT        (1)
+
+/*******************************/
+/* PS */
+/*******************************/
+/* Protected Storage (PS) Service definitions
+ * Note: Further documentation of these definitions can be found in the
+ * TF-M PS Integration Guide.
+ */
+#define TFM_HAL_PS_FLASH_DRIVER FLASH_DEV_NAME_SE_SECURE_FLASH
+
+/* In this target the CMSIS driver requires only the offset from the base
+ * address instead of the full memory address.
+ */
+/* Base address of dedicated flash area for PS */
+#define TFM_HAL_PS_FLASH_AREA_ADDR      HOST_ACCESS_PS_BASE_OFFSET
+/* Size of dedicated flash area for PS */
+#define TFM_HAL_PS_FLASH_AREA_SIZE      HOST_FLASH0_PS_SIZE
+/* Number of physical erase sectors per logical FS block */
+#define TFM_HAL_PS_SECTORS_PER_BLOCK    (4)
+/* Smallest flash programmable unit in bytes */
+#define TFM_HAL_PS_PROGRAM_UNIT         (1)
+#endif /* TFM_PARTITION_PROTECTED_STORAGE */
+
+
 #endif /* __FLASH_LAYOUT_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/host_base_address.h b/platform/ext/target/arm/rss/kronos/host_base_address.h
index a37a531fd..022914f8c 100644
--- a/platform/ext/target/arm/rss/kronos/host_base_address.h
+++ b/platform/ext/target/arm/rss/kronos/host_base_address.h
@@ -256,13 +256,23 @@
 #define HOST_AP_RSS_MAILBOX_ATU_SIZE   ALIGN_UP(0x1000, RSS_ATU_PAGE_SIZE)
 #define HOST_AP_RSS_MAILBOX_ATU_ID     RSS_ATU_AP_RSS_MAILBOX_ATU_ID
 
-
-/* Safety Island NVM flash */
-#define SI_FLASH_BASE           0x60000000   /* Safety Island NVM flash base address */
-#define SI_FLASH_SIZE           0x8000000    /* 128 MB */
+/* Safety Island NVM flash layout */
+/*
+ * ┌─────────────────────┬──────────────────────┐
+ * │ Signed Images       │  Protected Storage   │
+ * │     (64MB)          │       (64KB)         │
+ * └─────────────────────┴──────────────────────┘
+ *
+ */
+/* Safety Island NVM flash base address */
+#define SI_FLASH_BASE           0x60000000
+/* This part is for signed images. 64MB */
+#define SI_FLASH_IMG_SIZE       0x4000000
+/* This part is for Protected Storage. 64KB */
+#define SI_FLASH_PS_SIZE        0x10000
 
 /* Safety Island NVM flash logical base address using Non-secure ATU region */
-#define SI_FLASH_BASE_NS_LOG    (HOST_ACCESS_BASE_NS + 0x8000000)
+#define SI_FLASH_BASE_NS_LOG    HOST_ACCESS_BASE_NS
 /* Safety Island NVM flash physical base address */
 #define SI_FLASH_BASE_NS_PHY    (HOST_SI_PHYS_BASE + SI_FLASH_BASE)
 
@@ -275,4 +285,15 @@
 #define HOST_SI_CL1_SRAM_PHYS_BASE      (HOST_SI_PHYS_BASE + SI_CL1_SRAM_BASE)
 #define HOST_SI_CL2_SRAM_PHYS_BASE      (HOST_SI_PHYS_BASE + SI_CL2_SRAM_BASE)
 
+#ifdef TFM_PARTITION_PROTECTED_STORAGE
+/* Region into which to map Protected Storage */
+#define HOST_ACCESS_PS_BASE_S           (SI_FLASH_BASE_NS_LOG + SI_FLASH_IMG_SIZE)
+/* The offset is 64MB in SI Flash for PS */
+#define HOST_ACCESS_PS_BASE_OFFSET      SI_FLASH_IMG_SIZE
+
+/* The physical region base for PS */
+#define HOST_FLASH0_PS_BASE             (SI_FLASH_BASE_NS_PHY + SI_FLASH_IMG_SIZE)
+#define HOST_FLASH0_PS_SIZE             SI_FLASH_PS_SIZE
+#endif /* TFM_PARTITION_PROTECTED_STORAGE */
+
 #endif  /* __HOST_BASE_ADDRESS_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/host_drivers/cfi_drv.c b/platform/ext/target/arm/rss/kronos/host_drivers/cfi_drv.c
new file mode 100644
index 000000000..b7e1495d4
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/host_drivers/cfi_drv.c
@@ -0,0 +1,312 @@
+/*
+ * Copyright (c) 2015-2021, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+
+#include "cfi_drv.h"
+
+/*
+ * This memory is organized as an interleaved memory of two chips
+ * with a 16 bit word. It means that every 32 bit access is going
+ * to access to two different chips. This is very important when
+ * we send commands or read status of the chips.
+ */
+
+/*
+ * DWS ready poll retries. The number of retries in this driver have been
+ * obtained empirically from Juno. FVP implements a zero wait state NOR flash
+ * model
+ */
+#define DWS_WORD_PROGRAM_RETRIES	1000
+#define DWS_WORD_ERASE_RETRIES		3000000
+#define DWS_WORD_LOCK_RETRIES		1000
+
+/* Helper macro to detect end of command */
+#define NOR_CMD_END (NOR_DWS | (NOR_DWS << 16l))
+#define NOR_CMD_END_8BIT (NOR_DWS)
+
+
+/* Helper macros to access two flash banks in parallel */
+#define NOR_2X16(d)			((d << 16) | (d & 0xffff))
+
+static inline void mmio_write_32(uintptr_t addr, uint32_t value)
+{
+	*(volatile uint32_t*)addr = value;
+}
+
+static inline void mmio_write_8(uintptr_t addr, uint8_t value)
+{
+	*(volatile uint8_t*)addr = value;
+}
+
+static inline uint32_t mmio_read_32(uintptr_t addr)
+{
+	return *(volatile uint32_t*)addr;
+}
+
+uint8_t nor_cfi_reg_read(uintptr_t addr){
+	return *(volatile uint8_t*)addr;
+}
+
+void nor_cfi_reg_write(uintptr_t addr, uint32_t value){
+	*(volatile uint32_t*)addr = value;
+}
+
+static unsigned int nor_status(uintptr_t base_addr)
+{
+	unsigned long status;
+
+	nor_send_cmd(base_addr, NOR_CMD_READ_STATUS_REG);
+	status = mmio_read_32(base_addr);
+	status |= status >> 16; /* merge status from both flash banks */
+
+	return status & 0xFFFF;
+}
+
+void nor_send_cmd(uintptr_t base_addr, unsigned long cmd)
+{
+	mmio_write_32(base_addr, NOR_2X16(cmd));
+}
+
+void nor_send_cmd_byte(uintptr_t base_addr, unsigned long cmd)
+{
+	mmio_write_8(base_addr, cmd);
+}
+
+/*
+ * Poll Write State Machine.
+ * Return values:
+ *    0      = WSM ready
+ *    -EBUSY = WSM busy after the number of retries
+ */
+static enum cfi_error_t nor_poll_dws(uintptr_t base_addr, unsigned long int retries)
+{
+	unsigned long status;
+
+	do {
+		nor_send_cmd(base_addr, NOR_CMD_READ_STATUS_REG);
+		status = mmio_read_32(base_addr);
+		if ((status & NOR_CMD_END) == NOR_CMD_END)
+			return 0;
+	} while (retries-- > 0);
+
+	return -CFI_ERR_DEV_BUSY;
+}
+
+/*
+ * Poll Write State Machine.
+ * Return values:
+ *    0      = WSM ready
+ *    -EBUSY = WSM busy after the number of retries
+ */
+static enum cfi_error_t nor_poll_dws_byte(uintptr_t base_addr, unsigned long int retries)
+{
+	unsigned long status;
+
+	do {
+		nor_send_cmd_byte(base_addr, NOR_CMD_READ_STATUS_REG);
+		status = nor_cfi_reg_read(base_addr);
+		if ((status & NOR_CMD_END_8BIT) == NOR_CMD_END_8BIT)
+			return 0;
+	} while (retries-- > 0);
+
+	return -CFI_ERR_DEV_BUSY;
+}
+
+/*
+ * Return values:
+ *    0      = success
+ *    -EPERM = Device protected or Block locked
+ *    -EIO   = General I/O error
+ */
+static enum cfi_error_t nor_full_status_check(uintptr_t base_addr)
+{
+	unsigned long status;
+
+	/* Full status check */
+	status = nor_status(base_addr);
+
+	if (status & (NOR_PS | NOR_BLS | NOR_ESS | NOR_PSS))
+		return -CFI_ERR_DEV_PROTECTED;
+	if (status & (NOR_VPPS | NOR_ES))
+		return -CFI_ERR_GENERAL_IO;
+	return 0;
+}
+
+
+/*
+ * This function programs a word in the flash. Be aware that it only
+ * can reset bits that were previously set. It cannot set bits that
+ * were previously reset. The resulting bits = old_bits & new bits.
+ * Return values:
+ *  0 = success
+ *  otherwise it returns a negative value
+ */
+enum cfi_error_t nor_word_program(uintptr_t base_addr, unsigned long data)
+{
+	uint32_t status;
+	int ret;
+
+	nor_send_cmd(base_addr, NOR_CMD_CLEAR_STATUS_REG);
+
+	/* Set the device in write word mode */
+	nor_send_cmd(base_addr, NOR_CMD_WORD_PROGRAM);
+	mmio_write_32(base_addr, data);
+
+	ret = nor_poll_dws(base_addr, DWS_WORD_PROGRAM_RETRIES);
+	if (ret == 0) {
+		/* Full status check */
+		nor_send_cmd(base_addr, NOR_CMD_READ_STATUS_REG);
+		status = mmio_read_32(base_addr);
+
+		if (status & (NOR_PS | NOR_BLS)) {
+			nor_send_cmd(base_addr, NOR_CMD_CLEAR_STATUS_REG);
+			ret = -CFI_ERR_DEV_PROTECTED;
+		}
+	}
+
+	if (ret == 0)
+		ret = nor_full_status_check(base_addr);
+	nor_send_cmd(base_addr, NOR_CMD_READ_ARRAY);
+
+	return ret;
+}
+
+
+enum cfi_error_t nor_byte_program(uintptr_t base_addr, uint8_t data)
+{
+	uint32_t status;
+	int ret;
+
+	nor_send_cmd_byte(base_addr, NOR_CMD_CLEAR_STATUS_REG);
+
+	/* Set the device in write byte mode */
+	nor_send_cmd_byte(base_addr, NOR_CMD_WORD_PROGRAM);
+	mmio_write_8(base_addr, data);
+
+	ret = nor_poll_dws_byte(base_addr, DWS_WORD_PROGRAM_RETRIES);
+	if (ret == 0) {
+		/* Full status check */
+		nor_send_cmd_byte(base_addr, NOR_CMD_READ_STATUS_REG);
+		status = nor_cfi_reg_read(base_addr);
+
+		if (status & (NOR_PS | NOR_BLS)) {
+			nor_send_cmd_byte(base_addr, NOR_CMD_CLEAR_STATUS_REG);
+			ret = -CFI_ERR_DEV_PROTECTED;
+		}
+	}
+
+
+	nor_send_cmd_byte(base_addr, NOR_CMD_READ_ARRAY);
+
+	return ret;
+}
+
+/*
+ * Erase a full 256K block
+ * Return values:
+ *  0 = success
+ *  otherwise it returns a negative value
+ */
+enum cfi_error_t nor_erase(uintptr_t base_addr)
+{
+	int ret;
+
+	nor_send_cmd(base_addr, NOR_CMD_CLEAR_STATUS_REG);
+
+	nor_send_cmd(base_addr, NOR_CMD_BLOCK_ERASE);
+	nor_send_cmd(base_addr, NOR_CMD_BLOCK_ERASE_ACK);
+
+	ret = nor_poll_dws(base_addr, DWS_WORD_ERASE_RETRIES);
+	if (ret == 0)
+		ret = nor_full_status_check(base_addr);
+	nor_send_cmd(base_addr, NOR_CMD_READ_ARRAY);
+
+	return ret;
+}
+
+enum cfi_error_t nor_erase_byte(uintptr_t base_addr)
+{
+	enum cfi_error_t ret;
+
+	nor_send_cmd_byte(base_addr, NOR_CMD_CLEAR_STATUS_REG);
+
+	nor_send_cmd_byte(base_addr, NOR_CMD_BLOCK_ERASE);
+	nor_send_cmd_byte(base_addr, NOR_CMD_BLOCK_ERASE_ACK);
+
+	ret = nor_poll_dws_byte(base_addr, DWS_WORD_ERASE_RETRIES);
+
+	nor_send_cmd_byte(base_addr, NOR_CMD_READ_ARRAY);
+
+	return ret;
+}
+
+/*
+ * Lock a full 256 block
+ * Return values:
+ *  0 = success
+ *  otherwise it returns a negative value
+ */
+enum cfi_error_t nor_lock(uintptr_t base_addr)
+{
+	enum cfi_error_t ret;
+
+	nor_send_cmd(base_addr, NOR_CMD_CLEAR_STATUS_REG);
+
+	nor_send_cmd(base_addr, NOR_CMD_LOCK_UNLOCK);
+	nor_send_cmd(base_addr, NOR_LOCK_BLOCK);
+
+	ret = nor_poll_dws(base_addr, DWS_WORD_LOCK_RETRIES);
+	if (ret == 0)
+		ret = nor_full_status_check(base_addr);
+	nor_send_cmd(base_addr, NOR_CMD_READ_ARRAY);
+
+	return ret;
+}
+
+/*
+ * unlock a full 256 block
+ * Return values:
+ *  0 = success
+ *  otherwise it returns a negative value
+ */
+enum cfi_error_t nor_unlock(uintptr_t base_addr)
+{
+	enum cfi_error_t ret;
+
+	nor_send_cmd(base_addr, NOR_CMD_CLEAR_STATUS_REG);
+
+	nor_send_cmd(base_addr, NOR_CMD_LOCK_UNLOCK);
+	nor_send_cmd(base_addr, NOR_UNLOCK_BLOCK);
+
+	ret = nor_poll_dws(base_addr, DWS_WORD_LOCK_RETRIES);
+	if (ret == 0)
+		ret = nor_full_status_check(base_addr);
+	nor_send_cmd(base_addr, NOR_CMD_READ_ARRAY);
+
+	return ret;
+}
+
+
+
+unsigned int nor_id_check(uintptr_t base_addr)
+{
+	unsigned long status,status1,status2;
+
+	nor_send_cmd(base_addr, NOR_CMD_READ_ID_CODE);
+	status = mmio_read_32(base_addr);
+	status1 = mmio_read_32(base_addr+1*4);
+	status2 = mmio_read_32(base_addr+2*4);
+
+	CFI_FLASH_LOG_MSG("STATUS ID: %X \n",status);
+	CFI_FLASH_LOG_MSG("STATUS ID: %X \n",status1);
+	CFI_FLASH_LOG_MSG("STATUS ID: %X \n",status2);
+	status |= status >> 16; /* merge status from both flash banks */
+
+	CFI_FLASH_LOG_MSG("STATUS ID: %X \n",status);
+
+	return status & 0xFFFF;
+}
diff --git a/platform/ext/target/arm/rss/kronos/host_drivers/cfi_drv.h b/platform/ext/target/arm/rss/kronos/host_drivers/cfi_drv.h
new file mode 100644
index 000000000..75e93ad70
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/host_drivers/cfi_drv.h
@@ -0,0 +1,85 @@
+/*
+ *
+ * Copyright (c) 2021, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef CFI_DRV_H
+#define CFI_DRV_H
+
+#include <stdint.h>
+#include <stdbool.h>
+
+/* First bus cycle */
+#define NOR_CMD_READ_ARRAY		0xFF
+#define NOR_CMD_READ_ID_CODE		0x90
+#define NOR_CMD_READ_QUERY		0x98
+#define NOR_CMD_READ_STATUS_REG		0x70
+#define NOR_CMD_CLEAR_STATUS_REG	0x50
+#define NOR_CMD_WRITE_TO_BUFFER		0xE8
+#define NOR_CMD_WORD_PROGRAM		0x40
+#define NOR_CMD_BLOCK_ERASE		0x20
+#define NOR_CMD_LOCK_UNLOCK		0x60
+#define NOR_CMD_BLOCK_ERASE_ACK		0xD0
+
+/* Second bus cycle */
+#define NOR_LOCK_BLOCK			0x01
+#define NOR_UNLOCK_BLOCK		0xD0
+
+/* Status register bits */
+#define NOR_DWS				(1 << 7)
+#define NOR_ESS				(1 << 6)
+#define NOR_ES				(1 << 5)
+#define NOR_PS				(1 << 4)
+#define NOR_VPPS			(1 << 3)
+#define NOR_PSS				(1 << 2)
+#define NOR_BLS				(1 << 1)
+#define NOR_BWS				(1 << 0)
+
+/* Set 1 to enable debug messages */
+#define DEBUG_CFI_FLASH         0
+
+#include <stdio.h>
+#if (DEBUG_CFI_FLASH == 1)
+    #define CFI_FLASH_LOG_MSG(f_, ...) printf((f_), ##__VA_ARGS__)
+#else
+    #define CFI_FLASH_LOG_MSG(f_, ...)
+#endif
+
+enum cfi_error_t {
+    CFI_ERR_NONE,
+    CFI_ERR_WRONG_ARGUMENT,
+    CFI_ERR_NOT_INITIALIZED,
+    CFI_ERR_DEV_BUSY,
+    CFI_ERR_GENERAL_IO,
+    CFI_ERR_DEV_PROTECTED
+};
+
+/**
+ * \brief CFI device configuration structure
+ */
+struct cfi_dev_cfg_t {
+    const uint32_t base; /*!< CFI base address */
+};
+
+/**
+ * \brief CFI controller device structure
+ */
+struct cfi_dev_t {
+    const struct cfi_dev_cfg_t* const cfg;
+};
+
+/* Public API */
+void nor_send_cmd(uintptr_t base_addr, unsigned long cmd);
+void nor_send_cmd_byte(uintptr_t base_addr, unsigned long cmd);
+enum cfi_error_t nor_word_program(uintptr_t base_addr, unsigned long data);
+enum cfi_error_t nor_byte_program(uintptr_t base_addr, uint8_t data);
+enum cfi_error_t nor_lock(uintptr_t base_addr);
+enum cfi_error_t nor_unlock(uintptr_t base_addr);
+enum cfi_error_t nor_erase(uintptr_t base_addr);
+unsigned int nor_id_check(uintptr_t base_addr);
+uint8_t nor_cfi_reg_read(uintptr_t addr);
+void nor_cfi_reg_write(uintptr_t addr, uint32_t value);
+#endif
diff --git a/platform/ext/target/arm/rss/kronos/rss_comms/rss_comms_permissions_hal.c b/platform/ext/target/arm/rss/kronos/rss_comms/rss_comms_permissions_hal.c
index 41e9cf49e..d198d1229 100644
--- a/platform/ext/target/arm/rss/kronos/rss_comms/rss_comms_permissions_hal.c
+++ b/platform/ext/target/arm/rss/kronos/rss_comms/rss_comms_permissions_hal.c
@@ -25,6 +25,10 @@
 #ifdef TFM_PARTITION_PLATFORM
 #include "tfm_platform_api.h"
 #endif /* TFM_PARTITION_PLATFORM */
+#ifdef TFM_PARTITION_PROTECTED_STORAGE
+#include "tfm_ps_defs.h"
+#endif /* TFM_PARTITION_PROTECTED_STORAGE */
+
 
 #define INVALID_REGION_COUNTER_MAX  128
 #define INVALID_SERVICE_COUNTER_MAX 64
@@ -150,6 +154,20 @@ enum tfm_plat_err_t comms_permissions_service_check(psa_handle_t handle,
             goto out_err;
         }
 #endif /* TFM_PARTITION_PLATFORM */
+#ifdef TFM_PARTITION_PROTECTED_STORAGE
+    case TFM_PROTECTED_STORAGE_SERVICE_HANDLE:
+        switch(type) {
+        case TFM_PS_SET:
+        case TFM_PS_GET:
+        case TFM_PS_GET_INFO:
+        case TFM_PS_REMOVE:
+        case TFM_PS_GET_SUPPORT:
+            return TFM_PLAT_ERR_SUCCESS;
+        default:
+            goto out_err;
+        }
+#endif /* TFM_PARTITION_PROTECTED_STORAGE */
+
     default:
         goto out_err;
     }
-- 
2.25.1

